### ISSUE TITLE
Persona 5: Random font/texture swapping

### ISSUE THREAD

**Issue by UnaKaya** [RPCS3/rpcs3#4298]
Hello RPCS3 Team!

First off, let me just say a million thanks to the dev team. I wouldn't have been able to play this if it weren't for you guys.

I'd like to report an issue regarding Persona 5. This usually happens after playing the game for a period of time. It's the worst when my status bar isn't displayed correctly. I'm not sure if this is a normal thing but it never did this before.

Here are my specs:
CPU: i5-7600K
GPU: GTX 1060 6GB
OS: Windows 10
RPCS3: 0.0.4-6439-ee88e7f94

Sorry, totally new to this. I'm attaching censored screenshots of the issue and as well as the log file. I censored some details to mostly keep it spoiler-free:

![rpcs3_p5visbug2](https://user-images.githubusercontent.com/37035624/37557383-cabc90a4-2a3e-11e8-9199-7b95f38f5d99.jpg)
![rpcs3_p5visbug1](https://user-images.githubusercontent.com/37035624/37557384-cf3bee5e-2a3e-11e8-9bc4-45b8e6a31295.jpg)
![rpcs3_p5visbug4](https://user-images.githubusercontent.com/37035624/37557386-d097f176-2a3e-11e8-9d58-19f51d1c0f82.jpg)

Log link: http://www.mediafire.com/file/23u46l5p1qewoxa/RPCS3.log.gz

**Comment by Drayano** [comment]
It's a pain to find out which version broke the game that much since the bug happen randomly and I haven't found a way to reproduce it yet, unless playing for extended periods of time. There is an old build i was using where i'm 100% sure the bug doesn't happen which is the 0.0.4-6307-5959411ae if someone feels like bisecting which build introduced the regression.

**Comment by UnaKaya** [comment]
I found that exploring the Metaverse triggers the graphic glitches much faster than when going through Palaces.

**Comment by UnaKaya** [comment]
BTW, this was the build I used initially. 

https://ci.appveyor.com/project/kd-11/rpcs3/build/0.0.2-783

**Comment by kd-11** [comment]
That makes this not an error then. That means another thread (usually spu/mfc) has stalled and we have to make rsx "unstuck". Of course the once that happens its in an undefined state. The current stall detection timeout is 1 second (1000000 microseconds). You can make it infinite if you like in the config file under rsx hang detection timeout variable.

**Comment by UnaKaya** [comment]
What do you advise kd-11?

**Comment by UnaKaya** [comment]
I’m not so sure about the “semaphore_timeout” thing since it might just be a coincidence. What i’m sure of is that this happened when I downloaded later builds.

**Comment by kd-11** [comment]
Its very unlikely to be just a coincidence - its is a fairly serious issue whenever a cell core stalls and takes down the graphics chip. Try setting a much higher timeout (in older builds the waiting was infinite) like 30 seconds (30000000) in the yml configuration. Generally once you see an error like that, or the other fatal errors about a spurs kernel crashing, expect very bad things to start happening. I'd say, find the next save spot and restart the emulator. There are a few games with very broken emulation where the renderer ignoring the stalling in cell is not a big deal but those are few and its just a lucky coincidence.

**Comment by UnaKaya** [comment]
Alright, thanks for the support kd-11. I’m glad I reported this, hope it gets fixed!

Thanks a million!

**Comment by ruipgpinheiro** [comment]
I see this exact same issue. Upping the timeout to 30 seconds simply causes regular 30-second-long freezes, and still gives us corrupted textures. I tried disabling all SPU tweaks as well to no avail.

Reverting to rpcs3-v0.0.4-2018-02-12-95c6ac69 fixes the issue completely even with the default timeout and SPU tweaks enabled. This build does not have an unlimited timeout, so to me it looks like a RPCS3 regression causing one of the threads to stall. I've been using that older build to play through the game, but sadly it has lower performance than the more modern build.

As such, I'm not sure whether this should be closed?

**Comment by ruipgpinheiro** [comment]
@kd-11 
> That makes this not an error then.

I don't understand how you come to this conclusion - it is something that does not happen when running the game on an original PS3, and does not happen on 2-months-old builds. Could you explain in more detail why this is not a bug in RPCS3?

Do not take this wrong, I'm actually curious why you think this is not a RPCS3 bug. Is there any documentation on how to debug these kinds of bugs? I do CPU design as a profession, but even though I've considered in the past, I have never delved into an emulator before so was wondering how to best approach something like this (assuming it *is* a RPCS3 bug).

**Comment by MSuih** [comment]
>it is something that does not happen when running the game on an original PS3

IIRC this graphical glitch can actually occur on real PS3 as well, it's incredibly rare though because real hardware doesn't stall like RPCS3 does. It is not a bug with emulator but rather an unusual situation that occurs more frequently on rpcs3 and which the developers haven't prepared the game for.

The actual bug behind it is the other threads stalling and taking down rsx with it, and as KD said it affects many other games and not just Persona 5. The grapical glitches caused by it is a bug in the game code.

**Comment by ruipgpinheiro** [comment]
@MSuih
Thanks for the reply. I can understand that this issue (stalling bringing down rsx) could happen in the real console, but something must have changed to cause it to occur once every 10 minutes on average compared to a 2-months-old build where I haven't had it occur once in 8 hours of gameplay...

**Comment by ruipgpinheiro** [comment]
Just for further information, considering that rpcs3-v0.0.4-2018-02-12-95c6ac69 does not present the issue and 0.0.4-6439-ee88e7f94 does, we can easily conclude that whatever is causing stalls to occur often occured due to https://github.com/RPCS3/rpcs3/pull/4162.

**Comment by ruipgpinheiro** [comment]
I managed to build RPCS3 in my system, and started bisecting the pull request to try and find the exact commit that caused this issue.

First, I checked out ee88e7f94 and tested how reliably I could reproduce this bug:
Turns out that entering rooms in the first palace triggers this issue very often, so if I keep entering and leaving a single room I am able to reproduce this issue (the semaphore) in less than a minute every time I try.

With this easy way to reproduce the issue, I tried the following commits:

- 3406cc9 - Could reproduce in under 60s
- b67f28e - Could reproduce in under 60s
- 02e571a - Could *not* reproduce even after 10 minutes of constantly entering rooms
- 89c548b - Could reproduce in under 60s

As such, I can conclude that 89c548b is at fault.

**Comment by ruipgpinheiro** [comment]
> diff --git a/rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp b/rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp
> index 076a86b6e..825aa16f1 100644
> --- a/rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp
> +++ b/rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp
> @@ -810,12 +810,10 @@ std::string FragmentProgramDecompiler::Decompile()
>  				LOG_ERROR(RSX, "Unimplemented SIP instruction: CAL");
>  				break;
>  			case RSX_FP_OPCODE_FENCT:
> -				AddCode("//FENCT");
> -				forced_unit = FORCE_SCT;
> +				LOG_ERROR(RSX, "Unimplemented SIP instruction: FENCT");
>  				break;
>  			case RSX_FP_OPCODE_FENCB:
> -				AddCode("//FENCB");
> -				forced_unit = FORCE_SCB;
> +				LOG_ERROR(RSX, "Unimplemented SIP instruction: FENCB");
>  				break;
>  			case RSX_FP_OPCODE_IFE:
>  				AddCode("if($cond)");

Seems to resolve the issue, so it must be something related to the FENCT/FENCB instructions.

**Comment by ruipgpinheiro** [comment]
1h of gameplay without seeing any issue issue, I'm confident to say that the diff below (applied to 89c548b) "fixes" this issue when using OpenGL:

```
diff --git a/rpcs3/Emu/RSX/GL/GLTextureCache.h b/rpcs3/Emu/RSX/GL/GLTextureCache.h
index e35dc30e0..8cc70a3f9 100644
--- a/rpcs3/Emu/RSX/GL/GLTextureCache.h
+++ b/rpcs3/Emu/RSX/GL/GLTextureCache.h
@@ -812,28 +812,9 @@ namespace gl
 			if (context != rsx::texture_upload_context::blit_engine_dst)
 			{
 				cached.protect(utils::protection::ro);
-			}
-			else
-			{
-				//TODO: More tests on byte order
-				//ARGB8+native+unswizzled is confirmed with Dark Souls II character preview
-				if (gcm_format == CELL_GCM_TEXTURE_A8R8G8B8)
-				{
-					bool bgra = (flags == rsx::texture_create_flags::native_component_order);
-					cached.set_format(bgra? gl::texture::format::bgra : gl::texture::format::rgba, gl::texture::type::uint_8_8_8_8, false);
-				}
-				else
-				{
-					cached.set_format(gl::texture::format::rgb, gl::texture::type::ushort_5_6_5, true);
-				}
-
-				cached.make_flushable();
-				cached.set_dimensions(width, height, depth, (rsx_size / height));
-				cached.protect(utils::protection::no);
-				no_access_range = cached.get_min_max(no_access_range);
+				update_cache_tag();
 			}
 
-			update_cache_tag();
 			return &cached;
 		}
```

An equivalent (untested) workaround for Vulkan is:

```
--- a/rpcs3/Emu/RSX/VK/VKTextureCache.h
+++ b/rpcs3/Emu/RSX/VK/VKTextureCache.h
@@ -748,17 +748,9 @@ namespace vk
 			if (context != rsx::texture_upload_context::blit_engine_dst)
 			{
 				region.protect(utils::protection::ro);
-				read_only_range = region.get_min_max(read_only_range);
-			}
-			else
-			{
-				//TODO: Confirm byte swap patterns
-				region.protect(utils::protection::no);
-				region.set_unpack_swap_bytes(true);
-				no_access_range = region.get_min_max(no_access_range);
+				update_cache_tag();
 			}
 
-			update_cache_tag();
 			return &region;
 		}
 ```

I'm not sure what the underlying issue is, but it is certain that the above works around the issue and stops Persona 5 from deadlocking the RSX every 5-10 minutes of gameplay.

@kd-11 Any ideas?
@UnaKaya Could you re-open the issue?

**Comment by UnaKaya** [comment]
@ruipgpinheiro Done and done. Appreciate what you guys are doing! I've been reading your comments through e-mail notifications and honestly forgot that I haven't reopened the issue.

Tagging @kd-11

**Comment by ruipgpinheiro** [comment]
Just as a further update, I ported this workaround to the latest master and can confirm that I am unable to reproduce the issue even after 20 minutes of opening doors (using OpenGL. at least).

**Comment by kd-11** [comment]
Thanks for finding the trigger @ruipgpinheiro 
Unfortunately its not so simple as just disabling blit engine cache, and does affect the flow of cell memory read/write instructions indirectly. In fact for that blit engine resources exist but are not guarded so their contents are undefined. I'll look into it later, its kinda really hard to debug something that happens every 10 minutes or so since you never really know if its fixed...

**Comment by ruipgpinheiro** [comment]
@kd-11 No problem, thanks for your work!

I've been playing through Persona 5 so this issue was bothering me quite a bit. Downgrading to the old version fixed it, but I noticed that I lost some performance and there were missing effects so I wanted to figure out a workaround that gets me playing on master again. Obviously, it makes no sense to simply disable the blit engine cache for everything, but for my purposes running a custom build with it disabled for playing Persona 5 is acceptable until the underlying issue is fixed.

As for reproducing the bug, it seems to be able to occur at any point, but loading screens will trigger it most often. The easiest way I've found is to use AutoHotKey to hold down "S" (mapped to thumbstick down) and spam "X" (mapped to the X button). Then just go to the first in-game palace, place your back against a door, and activate the script. The game will then constantly open the same door (entering and leaving the room) multiple times per minute. 90% of the time, the issue occurs under a minute, although I've had at least one instance where it took around 10 minutes.

Stressing your computer with other stuff seems to trigger it more often as well - I would usually activate the script, wait a few minutes without anything occuring, and then when I open Youtube on another screen and start watching something suddenly I got the deadlock.

I also have the impression that OpenGL makes this deadlock occur more often, but I am not completely sure - I might just be imagining things.

**Comment by ruipgpinheiro** [comment]
In order to confirm the above suspicion that blit_engine_dst texture flushing is the thing triggering this deadlock, I tried the following code using Vulkan:

```diff --git a/rpcs3/Emu/RSX/VK/VKTextureCache.h b/rpcs3/Emu/RSX/VK/VKTextureCache.h
index e01d3756a..28b98e9f6 100644
--- a/rpcs3/Emu/RSX/VK/VKTextureCache.h
+++ b/rpcs3/Emu/RSX/VK/VKTextureCache.h
@@ -132,7 +132,7 @@ namespace vk
 		bool is_flushable() const
 		{
 			//This section is active and can be flushed to cpu
-			return (protection == utils::protection::no);
+			return (protection == utils::protection::no && context != rsx::texture_upload_context::blit_engine_dst);
 		}
 
 		bool is_flushed() const```

I can confirm that with the above diff applied, I have experienced no deadlocks on Vulkan in over 30 minutes.

**Comment by kd-11** [comment]
Well, that just disables flushing on blit engine derived sections.
The problem itself isn't graphical- its deeper than that. Protected memory sections are protected as their data exists on the host GPU and not the cpu where Cell is being emulated. When the game decides it want to read something that does not exist on the cpu, an access violation is raised - this is the tricky part. If the data does not exist, its forcefully copied back, a process that is fairly quick as there is a cheap prediction algorithm to anticipate this. However, it takes enough cycles to memcpy large memory areas (framebuffers and texture regions can easily consume upto 16M) to desynchronize threads. Due to lack of memory mirrors, there is also a brief moment where the memory is unprotected, but has invalid contents. i.e Picture 6 racing spu threads running some task in parallel modifying data existing on the GPU. One thread reaches the violation handler before the others and acquires the cache mutex. It unprotects the memory range to write the gpu-resident data on the cpu side. However, another thread reaches this section late due to host scheduling and finds the memory unprotected, so it proceeds to consume what is now undefined garbage since the first thread is still writing the data. Its an architectural flaw with no simple solution and it affects many other games in subtle ways that often seem to resolve themselves. An easy way to test this it to run something with interpreters only and then surround memory flush with a system pause/resume to block everything until the transfer is completed (terrible for performance and causes microstutter). Another title affected by obvious corruption is demons souls, but minimizing the undefined transfer window greatly reduced the number of occurrences.

**Comment by kd-11** [comment]
I should also mention - blit engine is really just a DMA engine, it does not necessarily transfer images/textures. Its often used to move program code around as well which is why any data its holding needs to be synchronized tightly. The dma transfer has no context around it making it impossible to determine what is being transfered beforehand.

**Comment by ruipgpinheiro** [comment]
Thank you for the explanation. So if I understand you correctly, you're saying that the semaphore freezes are likely due to one of the thread consuming an area of memory that another thread was busy copying back, causing it to lock up? That could certainly explain why the freeze occurs only with the blit engine (e.g. if it starts executing incomplete code in a DMA region while another thread is busy writing it).

However, I'm not sure I understood the whole issue. The emulator seems to be aware of the on-going flush, so it could acquire a per-region mutex and prevent any further accesses to that region while it is busy, to explictly prevent accesses to the invalid contents. This obviously would sacrifice some performance, but at least a whole system pause wouldn't be needed?

**Comment by kd-11** [comment]
The access does fail! This is by design. Rpcs3 implements its own segfault handlers for this reason, access->segfault->flush and make pages available->proceed. There is no way to know what a 'region' consists of, pages can be shared between many things for example but the page will still be marked no-access if its data is not present. The memory protection is the region lock and it is not manageable with mutexes outside of guaranteed execution paths (the violation handler is mutex handled in the cache, but that means nothing if the thread does not hit the violation). We need memory mirrors, but for obvious security reasons no OS will let you do this easily.
As for p5, the fact that it is indeed reading memory that doesn't exist means there is something wrong with it disabled (it will read something as all 0) but its likely too subtle to notice. The same was true before some memory respecification code was made available - most users only noticed the difference when it was working properly. There is also the possibility that the fault is not caused by reading garbage but rather a partial page with shared blit resource memory and some other code getting corrupted by the flush. Hopefully its the second simpler case as the first one is really hard to fix properly without performance loss. I have also noticed the issue has seemingly gotten worse recently - no timeouts I think, just textures getting randomly garbled so it is something to look into soon.

**Comment by kd-11** [comment]
My post did little to answer your questions so I should maybe clarify a bit:
1. No, the emulator hangs because a cell thread does not follow the intended path and either another thread commands it to abort what its doing or it writes wrong data to the destination. Thats how nvidia GPUs are synchronized with 'semaphores'.
2. As for how disabling blit engine flush fixes the problem - the access violation has no writes just unprotect and proceed. Its much much faster because of this so the thread will keep running. I can write a simple test case to prove what is the cause of the situation. Just add a busy loop or sleep when flushing a blit engine resource, after unprotect but before violation handler returns.

**Comment by ruipgpinheiro** [comment]
Thank you very much for your time and explanation. It makes all makes much more sense now. I've actually had a few instances of full-on deadlocks when using OpenGL without the semaphore timing out, but I just assumed I had broken something while testing, I guess it wasn't a coincidence.

I'm not sure if there's anything else I can do at this point to help - I'm a computer architecture / hardware engineer guy, not a software engineer, so while I understand what the issue is, it's a bit out of my league to fix! Let me know if there's anything else I can do.

I guess I'll just force CPU blit (with WCB disabled) while playing this game. I'll have to see what kind of impact that has on performance - worst case, I'll just use my workaround above instead!

You mentioned before that other games have corruption because of similar reasons. Am I incorrect that these games also have no easy way to fix the issue without rewriting the whole virtual memory management system? Because from what you say, if the emulator wants full compatibility then that will be necessary at some point, right?

**Comment by kd-11** [comment]
Yes, these other titles also are not fixable until we can map one physical page to more than one virtual address from user space. There are ways to make it work on windows but you're basically fighting the OS by doing so. Its not a simple task but something has to be done to ensure the operation is completely hidden from the application. Do note that using cpu blit without wcb will likely remove some stuff like the semitransparent backgrounds during pause menu and the transition between loading screens so you can use your patch in the meantime.

**Comment by ruipgpinheiro** [comment]
PS: You've probably already thought of this, so I'm probably wrong, but wouldn't it be possible to do double-buffering on the memory pages? For example, do not unprotect the old page, just allocate a new page, flush the data to it, and then remap the new page to the old page's address. Once that is done, de-allocate the old page. Even better, if extra memory usage is acceptable, simply allocate 2 pages from the start, and every time you flush you write to the inactive one and then once it finishes, activate it.

**Comment by kd-11** [comment]
That has the same problem as the other solution - you cannot freely manipulate the paging tables from userspace. No OS will just give you that kind of access.

**Comment by ruipgpinheiro** [comment]
Hmm, off the top of my head, doesn't windows support multiple memory views of the same physical address by requesting a file mapping with an invalid handle? [CreateFileMapping](https://msdn.microsoft.com/en-gb/library/windows/desktop/aa366537(v=vs.85).aspx) seems to explictly state this supported:

>If hFile is INVALID_HANDLE_VALUE, the calling process must also specify a size for the file mapping object in the dwMaximumSizeHigh and dwMaximumSizeLow parameters. In this scenario, CreateFileMapping creates a file mapping object of a specified size that is backed by the system paging file instead of by a file in the file system.

Then you can map that page somewhere else using [MapViewOfFileEx](https://msdn.microsoft.com/en-us/library/windows/desktop/aa366763(v=vs.85).aspx), no? I thought mmap on linux had something similar.

Anyway, I digress. I trust that you know what you are doing, so my suggestions are probably just due to my lack of domain knowledge. Thank you very much for your help, I'll use my workaround patch above to keep playing. Let me know if there's anything I can do to help.

**Comment by kd-11** [comment]
Thats what I meant when I said the current memory management code would need to be written wholly to only work with file mappings. I'm also unsure if *nix OSes support this kind of feature in a straightforward manner. We currently use VirtualAlloc/VirtualFree/VirtualProtect to manage the virtual memory for the emulated ps3 system.

**Comment by ruipgpinheiro** [comment]
[This article](http://nullprogram.com/blog/2016/04/10/) seems to discuss the above approach on Windows and a similar approach on POSIX systems (although it doesn't seem to be completely supported).

**Comment by kd-11** [comment]
I remember having a long conversation about this very issue with other developers when I first figured it out. I cannot remember the counter-points to it, but I do remember all these options were on the table back then. Maybe the emulator has come a long way enough that this is now easier to do.

**Comment by ruipgpinheiro** [comment]
So, I've spent some time actually playing the game with the above workaround on the latest master using Vulkan, and have some updates to this bug:

1. While the semaphore timeout didn't occur during loading screens anymore, there is one location in-game where it occurs all the time. If I travel to the "Station Square" location, either manually of because of cutscenes, I get a dealock every 5-10 seconds even without loading screens.

2. Compared to 95c6ac69, there is a lot of "new" texture corruption on scene transitions. For example, when the loading screen shows up with people walking, the black background will sometimes be colored randomly. These "random textures" will show up very often, mainly in scene transition animations, but also in the menu and anywhere where there is UI drawn on top of a static image of the normal graphics.

Regarding 1., I'm not sure what could be causing it. I've just started avoiding Station Square as much as possible. Regarding 2., I bet that these animations/UI use the blit engine, and all that's happening is the workaround is making it show uninitialized memory. I'm curious as to why 95c6ac69 did not have these issues.

Nevertheless, the performance increased and graphics improvements (extra effects) are important enough to me that I'll continue playing with the workaround. Just thought I should update this bug with this new information.

**Comment by ruipgpinheiro** [comment]
@kd-11 That is great to hear! Thanks for your work, can't wait to have these issues resolved.

**Comment by kd-11** [comment]
Thanks for confirming. Will add a proper fix before merging the wip pr.

**Comment by Seafra01** [comment]
I am using a customized version based on 0.0.5-6675(The only edit is enabled TSX for Haswell CPUs). And I found something interesting, it seems as long as I press 'triangle' button which opens the menu in game, the texture glitch could be triggered very likely. Or not, no matter how many locations I teleport, how long I have played, the graphic is still good. Seems there is a connection between them, right? I think this might be helpful. So I decided to post here. Sorry for interrupting your dialogues.

**Comment by ruipgpinheiro** [comment]
@Seafra01 I have also noticed recently that opening/closing the menu repeatedly triggers the issue very quickly. I believe that doing anything that involves the blit engine can trigger the issue, and P5 seems to use it a lot for the UI transitions (loading screen start/stop, opening/closing menus, entering/exiting submenus, etc).

**Comment by kd-11** [comment]
https://github.com/RPCS3/rpcs3/pull/4505 should fix this issue. Note that memory mirrors have a quirk that is making it possible to reach a deadlock in some race conditions, but it does not seem to be hit by this game.

**Comment by ruipgpinheiro** [comment]
Thank you very much for letting us know. Next time I play I'll let you know if I see any texture corruption at all.

From reading that pull request, am I right that memory mirrors seem to have caused a bunch of regressions? People seem to be reporting that the blit engine is now broken in many games...

**Comment by kd-11** [comment]
Its resolved. Remaining issue would need a very lengthy explanation regarding how vm locks work in rpcs3 and how violation detours can lead to deadlock. Most games are now unlikely to trigger this deadlock though.

**Comment by ruipgpinheiro** [comment]
@kd-11 I just tested 6f1c67ed3b619e2060c35b1bfd76ecb2542204a2 and got a deadlock in the first 20 seconds of opening/closing the menu...

**Comment by UnaKaya** [comment]
Hello everyone! I'm no computer expert so I'm afraid I can't contribute more to the code...but I really appreciate what you guys are doing. Please do let me know if I could be of help somehow. I feel like Mishima right now haha

Anyways, gonna support @ruipgpinheiro on this one. Tried #4505 but the graphic glitch still happened

![untitled](https://user-images.githubusercontent.com/37035624/39957474-24730b6a-5626-11e8-96df-981f5cf40537.jpg)

**Comment by kd-11** [comment]
@UnaKaya that is a different glitch. The glitch in question isnt the menu one, its the 3d objects in the world getting corrupted. Things like the personas themselves or random objects in the game world. This error is partially related but its not going away soon (requires memory merge support which is nontrivial and waay too much effort right now for minimal gains).

**Comment by UnaKaya** [comment]
Ohh, I see. But if we're talking about 3D objects getting corrupted, yep it still happens. I battled a certain boss and Ann's head was a blue mess from the Matrix, Sandalphon's wings were fuzzy, etc.

I'll take a screenshot when it happens again. Quit playing P5 for a moment since it hanged haha!

**Comment by ruipgpinheiro** [comment]
30 mins of opening menus later, [this diff](https://github.com/RPCS3/rpcs3/files/1997900/workaround_6f1c67e.txt) seems to resolve the deadlock when applied to 6f1c67e.
I basically reset parts of the `trampled_set` loop in the `invalidate_range_impl_base` function to before your PR, and then disabled deferred_flush (simply disabling deferred_flush on your PR did *not* work).

I have no idea what the underlying cause might be, though. I simply spent some time trying to apply the changes of my previous "hackish fix" to your PR until something stopped dead-locking.

**Comment by ruipgpinheiro** [comment]
I've now played about 5h with 6f1c67e and the diff linked in my comment above, and can confirm it is perfect. I did not encounter a single deadlock or corrupted texture in all that time.

The game is not yet perfect; there's the menu texture glitches (is there an issue open for that?) and the the distortion glitches (#4086) still left to fix, but the deadlock/corrupted 3D textures bug we've been discussing is gone.

**Comment by kd-11** [comment]
That patch disables blit engine sync completely (its what deferred flush is). Inline flush is rarely used and is only usually relevant if wcb is enabled.

**Comment by kd-11** [comment]
Mobile interface mistake, reopened.

That patch disables blit engine sync completely (its what deferred flush is). Inline flush is rarely used and is only usually relevant if wcb is enabled. I suspect the long delay caused by memory manipulation also plays a role. Unfortunately its not possible to completely freeze the emulator when synchronizing vm contents so it might persist for the near future. I'll keep checking but I'm having a really hard time reproducing the corruption making it hard to test. Ofc I can't run tests for more than a few minutes at a time.

**Comment by ruipgpinheiro** [comment]
With regards to reproducing, it should be enough to test against RSX deadlocks, since the corrupted textures seem to be a symptom of those. And testing against deadlocks is relatively simple, just open and close the menu 100+ times (I've never had to wait more than 5 minutes for a deadlock while doing that). Use the AHK script above and you don't even need to have RPCS3 in focus.

**Comment by kd-11** [comment]
They really arent the same bug but they are related. I know exactly how both happen and the lockup is not the same as the color issue but they are closely related.  I just need simple way to reproduce so I can properly test and fix the issue. I guess I'll just have to fix it theoretically and hope it works.

**Comment by ruipgpinheiro** [comment]
(PS: Seems Persona 5 is broken on the latest master #4556, I get the same issue)

**Comment by kd-11** [comment]
I removed the new protection code as it is still very unstable. It will take some time before it is ready.

**Comment by ruipgpinheiro** [comment]
The new failure does not seem to be related to the protection code, it seems to have been caused by one of the commits be5c18c..7cb4672 (which your PR was rebased into)

**Comment by ruipgpinheiro** [comment]
So, taking your last message into account, I played around with the workaround a bit. It seems the issue is also resolved if, in addition to disabling deferred flush, I force an inline flush (by doing `allow_flush = !discard_only;`.

[Diff](https://github.com/RPCS3/rpcs3/files/1998550/workaround_minimal_e189da3.txt) for https://github.com/RPCS3/rpcs3/issues/4298#issuecomment-388625308

**Comment by ruipgpinheiro** [comment]
I've been playing with https://github.com/kd-11/rpcs3/commit/04fff860bcffb5e00f6678a7b58ee61e704a80e9 (no workaround applied) and it is very playable with a driver timeout of 1sec. The dead-lock still occurs once in a while, but I have yet to see any corrupted 3D texture.

However, the UI textures are very buggy in that commit. I am not just talking about the "slightly corrupted menus" and such, on that commit the frequency of corrupted UI textures is the highest I've seen yet by a factor of two at least. This goes from corrupted UI backgrounds, corrupted mini-map, corrupted character faces, corrupted loading screens, etc.

I've also seen multiple times the "talking character face" be either the incorrect character, or have bits and pieces (e.g. the "moving mouth) of the incorrect character, something which I had not experienced before.

**Comment by ruipgpinheiro** [comment]
I think I have discovered a quicker way to reproduce the 3D texture corruption: Simply go into the Velvet Room, choose the Dyad Guillotine, and start randomly opening persona information using Square. It seems opening persona info while in the Velvet room makes corruption quite likely, and I can usually get the persona texture to become corrupted in a minute or two.

**Comment by Seafra01** [comment]
@ruipgpinheiro dude,could you share your latest workaround ,the no side-effect version,please? I am   
very interested.

**Comment by ruipgpinheiro** [comment]
@Seafra01 I've submitted the current workaround to https://github.com/ruipgpinheiro/rpcs3/tree/p5_workaround

Note that the workaround is hackish, I simply disable deferred flush on top of @kd-11's memory mirror changes (and I have to revert some changes to get it to work). I haven't experienced issues in Persona 5 yet (other than the corrupted menu textures and very rarely a corrupted 3D texture) but I cannot guarantee anything. I also assume it will not work with any other game other than P5.

**Comment by ruipgpinheiro** [comment]
@kd-11 I spent some time trying to figure out what was going on. My current work can be found in https://github.com/ruipgpinheiro/rpcs3/commit/c1da07f9fbe61f0281780f5c72751b3607cb519a

I have yet to see a single corrupt texture with the above commit (I still encounter dead-locks, though), and the test_cpu_range tests have yet to fail even once. While I did multiple changes (one I believe is one bug-fix), the thing that actually stopped texture corruption was to force the protection policy to protect_policy_full_range (at the cost of a bit of performance).

I'll continue playing with the above build and keep you posted.

**Comment by Metalcape** [comment]
@ruipgpinheiro I've been following this discussion for a while, from what I've seen this issue is very specific to P5, so wouldn't it be possible to add a separate option in RPCS3 to toggle your "hackish fix" on/off? At least this would allow people to play on master build without texture corruption and just turn it off when they play something else. This could be a nice workaround while we wait for a definitive fix. Unless kd-11 is already working on it, that is.

**Comment by kd-11** [comment]
@ruipgpinheiro Using https://github.com/kd-11/rpcs3/commit/7bf4722d7004c194f80892fe500a440a6cb296e1 I cannot trigger the bug you have shown in your screenshot. Note that the changes added in that commit are minimal (range fixup should not affect the selected textures) and therefore it should be that the previous commit should also have been fine. Could you test without your hack? I get the feeling it may be interfering somehow with the expected flow. I have tried dyad guillotine (select + square, exit, select next then press square...) and it works fine. Maybe only happens on some personas?

**Comment by kd-11** [comment]
After about 2 hours of testing I have been able to get corrupted texture (not wrong texture, just corrupted) to show up on a persona. Switching selection R1/L1 cleared it up immediately though. I have also seen one corrupted UI texture which also reverted itself after exiting/re-entering the menu so I would say it is now a very rare occurrence. There are other titles with more serious issues though with the changeset.

**Comment by ruipgpinheiro** [comment]
@Metalcape As far as I know, RPCS3 has a policy against game-specific hacks, which I agree with. The Persona 5 workaround is very hackish, it breaks a ton of internal stuff and IMO it is merely coincidence that Persona 5 doesn't simply crash. If we start adding these hackish game-specific fixes to the master branch, soon there will be more hackish fixes than real code, and it will just cause an even bigger headache. If you really want to keep playing, either grab rpcs3-v0.0.4-2018-02-12-95c6ac69 (which does not have this issue), or build the workaround.

**Comment by Metalcape** [comment]
@ruipgpinheiro Oh OK, I didn't know that. It's perfectly understandable. It's not really a problem for me anyway as on my old CPU the Farseer build still has better performance than master, it was just a suggestion for people who play on master.

**Comment by ruipgpinheiro** [comment]
@kd-11 Sorry, I was busy with other stuff so only had time to write a more complete answer now.

I'll double-check once I have some time with https://github.com/kd-11/rpcs3/commit/7bf4722d7004c194f80892fe500a440a6cb296e1, but AFAIR I was using a fresh (unmodified) checkout of https://github.com/kd-11/rpcs3/commit/6d785ab057c703d59a7b44032ab22e093be65a5e when I was reproducing the issues.

The easiest way I've found to reproduce the issues is quite specific:

1. Go to the Dyad Guillotine "Normal Fusion" menu where you can select your personas to fuse.
2. Select a random persona
3. Repeatedly (and very quickly) press Square+Circle ("z" + "c" in the default keyboard mapping) to open and close the persona information dialogue. Repeat +/- 10 times very quickly.
4. Switch to a different Persona, press Square.
5. Try another Persona, press Square.
6. Go back to the Persona from step 3, press Square.
7. Was one of the textures corrupted? Bug triggered! Otherwise, repeat starting from step 2

With the above, I was usually able to reproduce the bug very quickly, in one or two tries, I can record a video of this if you want.

**Comment by kd-11** [comment]
I was using your dyad method and could not reproduce, although about 30 mins of gameplay can trigger it. I'm moving on to the timing hypothesis since the flush does not seem to actually do anything (corruption / wrong data is still present even if I do memset(0) instead of writing gpu memory). As for range check failing - yea, thats expected. It means the game 'freed' the memory used by blit engine and tried to load a texture there using memcpy() where the memory occupied by the texture happens to be in the middle. The 'head' becomes overwritten which is detected by the range check and invalidates the section via unprotection since memcpy is an advancing operation. I guess I should also mark as dirty but it doesnt matter since unprotected sections are ignored.

**Comment by ruipgpinheiro** [comment]
Did you specifically do the "Square -> Circle" spam and then switch to a different Persona as I explained above? Simply opening random personas doesn't seem to trigger it as often, I suspect the UI transition is the trigger.

**Comment by kd-11** [comment]
Yes I did square-circle for every persona, likely just hardware differences. You should really join our discord as it makes communicating test results easier without sending out email notifications to the entire team.

**Comment by ruipgpinheiro** [comment]
It is not "for every persona", but instead very quickly for a single persona some 10+ times, and only *then* do you switch to a different persona.

Oh, I did not see there was a RPCS3 Discord. I'll join it later once I get home and have some more time to look into this.

**Comment by kd-11** [comment]
Practically fixed by https://github.com/kd-11/rpcs3/commit/c73addd58abf82b5ced444c1de0472d3ddb30002 (build here https://ci.appveyor.com/project/kd-11/rpcs3/build/0.0.0.6-6798). If able to reproduce, let me know before the code is cleaned up for a PR.

**Comment by ruipgpinheiro** [comment]
I tested https://github.com/kd-11/rpcs3/commit/618d6d205bbd170406f3ab694416654d2fa04b2a and can confirm I was unable to reproduce the texture corruption/swap using the Dyad method. I played for a while with no noticeable issues as well, so it looks like that indeed worked!

(The deadlock and menu corruption still occur, but we know that those are different issues)

**Comment by kungfulon** [comment]
I'm using kd-11/rpcs3@618d6d2, not getting model texture corruption anymore, but I encountered this weird map texture corruption: [https://imgur.com/icxtDGD](https://imgur.com/icxtDGD)

**Comment by Tymec** [comment]
I am using the kd-11/rpcs3@618d6d2 and I haven't gotten the texture corription on the models themselves, but I still get it on the UI

**Comment by isshininu** [comment]
Those texture corruptions are very similar to what i got playing DeS.
![660](https://user-images.githubusercontent.com/6324523/40590737-fe0078f0-6214-11e8-9831-caa8379b1d14.png)

**Comment by kd-11** [comment]
1. Des issue is related, but should now be solved as memory mirrors were implemented recently
2. Thanks to @ruipin testing hacks in texture cache, deadlock is likely fixed in https://ci.appveyor.com/project/kd-11/rpcs3/build/1518

**Comment by ruipin** [comment]
I'll test the new fix once I get home. Glad I could help figure out the issue (even if my final guess at the real issue was actually wrong), thanks for your work!

Also am curious whether this fixes the huge number of timeouts some other games used to see...

**Comment by SaulinPav** [comment]
I've played an hour or two in Persona 5 and haven't noticed neither models corruption nor UI corruption on the latest build. I think the bug is fixed now. Thank you!

**Comment by Tymec** [comment]
For some reason Persona 5 is unplayable on the latest build for me. I got around 30fps in most places and sometimes even 60fps on kd-11/rpcs3@618d6d2
CPU: i5 6600k 4.5GHz
GPU GTX 770

**Comment by AniLeo** [comment]
That's unrelated to this issue. If this issue is fixed, then it can be closed.

**Comment by AniLeo** [comment]
You're not on latest master. Update, then post RPCS3.log alongside the screenshot.

**Comment by Tymec** [comment]
I'm actually on NVIDIA drivers.

**Comment by Tymec** [comment]
Happened again after about 2 hours of playing the game on the latest master.
![prtscr capture](https://user-images.githubusercontent.com/2727109/42350409-bda903ca-80b0-11e8-8449-bbca585e1e2d.png)
![prtscr capture_2](https://user-images.githubusercontent.com/2727109/42350410-bf99793a-80b0-11e8-9d0b-3b2bbd238ebb.png)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2168371/RPCS3.log.gz)

**Comment by ruipin** [comment]
Thought I should document my investigation into the remaining issues that I am personally aware of and fall withing this github issue. Maybe someone else might find this useful and has some idea of what could be causing this.

**Issue**

For reference, the issues discussed in this ticket number that were not resolved by the semaphore timeout fix are the texture swaps. There are two types of issues I am aware of, although they are very likely related:

1. Randomly, game textures (often UI textures, sometimes characters, personas or even the environment objects) will "swap" with other objects and display an incorrect texture. The most common instances of this issue I've seen are the character portraits in dialogue showing parts of the wrong person, or persona textures being broken in the velvet room. Usually, reproducing these issues can be done in around 10-15 minutes of gameplay if you repeatedly trigger dialogue or the persona details UI inside the velvet room.

2. In addition, there is a bug that has been present apparently since years ago which has to do with the Persona 5 fonts. Sometimes (but very rarely), the letter displayed will be incorrect. E.g., "Rakukaja" (a persona skill) becomes "Rmkukmjk". Closing and re-opening the UI is usually enough to make it become correct again.

**Workaround 1**

A workaround for 1. as I have mentioned multiple times in the past is to enable full-range protection policy for the texture cache. This can be done e.g. in Vulkan by building from source and changing the line https://github.com/RPCS3/rpcs3/blob/745ed8331cfc3fc4791984af1e9b07a472128261/rpcs3/Emu/RSX/VK/VKTextureCache.h#L35

This workaround resolves 99% of the issues I have seen of (1) outside the Velvet room, and seems to make (2) even more rare. However, it is not a complete "fix" since I can still reproduce (1) in the Velvet room, and sometimes (2) still occurs.

A hypothesis as to why this workaround works is that the default protection policy (conservative) for performance reasons does not protect all the pages that contain a specific texture. For example, if we have a texture A, if | is a page boundary, and (XYZ) denotes a protected page, we might have AAA|(AAAA). Depending on what the game does, it might be able to overwrite the beginning portion of the texture without RPCS3 realizing it happened, e.g. BBA|(AAAA) where B is a new texture, as long as the first word of A and B match. Full range protection will guarantee that there are no unprotected pages containing a texture, i.e. we'd have (AAA)|(AAAA).

However, this cannot be the whole story since as stated above the issues are only partially fixed by this workaround.


**Breaking things**

While I was originally assisting kd-11 with investigating the semaphore deadlock issue, at some point I managed to break my version causing (2) to occur ridiculously often. At the time I was not aware (2) was an issue, so I just thought I might have done something bad and quickly reverted my changes.

Suspecting that I might actually have stumbled onto the issue back then (by accidentally making it worse), I thought I should try and break it the same way again. After some time looking into this, I realized that breaking that if using the previous workaround, doing so is extremely simple.

With https://github.com/ruipin/rpcs3/commit/d04954bbe84fce2ca7eea3c47c813a636fa052f4, I see that fonts almost immediately become messed up with swapped letters, e.g.:

<img width="396" alt="rpcs3_2018-08-11_18-28-02" src="https://user-images.githubusercontent.com/11169970/43994421-62ba3302-9d94-11e8-9b22-d4857269cf72.png">


**Workaround 2**

With the new suspicion that the texture swap issue might be related to some bug in `get_intersecting_set` (since breaking that method causes texture swaps to happen very often), I started analyzing this code in detail and came up with a second work-around  that seems to eliminate all texture swap issues from my testing. I've tried playing with this for a few hours and have yet to see any texture swap.

https://github.com/ruipin/rpcs3/commit/99e52da838d1f075cfff7e7c4b810f2024533d91

Now, this is not a proper fix - although I do believe I might have found two separate issues in the code (see commit description, it would be nice if others could double-check).

However, even with those possible issues fixed, the only thing that resolves the actual problem is to force RPCS3 to use full-range bounds checking inside `get_intersecting_set`, instead of using the protected range only.

**Workaround 3**

After running out of ideas, I started randomly breaking code again (favorite way to debug code you don't fully understand, IMO). This is when I remembered I had long ago noticed an inconsistency between the way `invalidate_range_impl_base` unprotects `sections_to_unprotect` when `deferred_flush==false` and the way `flush_all` invalidates the same data structure.

To be exact, `flush_all` makes sure to call `tex.set_dirty(true);` on all such sections before unprotecting them, however `invalidate_range_impl_base` does not. Just for the heck of it, I reverted `get_intersecting_set` to use only the protected range, and added the `set_dirty` call to `invalidate_range_impl_base`.

https://github.com/ruipin/rpcs3/commit/9406c875c4c5747093091e40dc9c655763e17755

Turns out that this seems to indeed resolve the texture swaps, at least from my testing. I'm not sure if this inconsistency is actually a bug (or whether it is simply working around a bug somewhere else), so I am not going to call it a "fix".

**Comment by dkambersky** [comment]
Speaking for the people silently subscribed to this issue: thank you @ruipin for your determination and continued work on fixing this and related stuff! (feel free to delete if too offtopic)

**Comment by ruipin** [comment]
I believe I have finally tracked down the bug causing the texture swaps. It was related to the `set_dirty(true)` call after all.

The latest commit in https://github.com/RPCS3/rpcs3/pull/4970 should fix it, but I need more testers to make sure this is indeed the last we see of this issue.

(Note that it is theoretically possible that texture swaps might happen on any game without strict mode being enabled, but they should be extremely rare (or the game extremely poorly coded).)

**Comment by ruipin** [comment]
As a further update, I have tested my PR #4970 for over 5 hours. The texture swap bug is indeed fixed at least during normal gameplay, and I never saw a font swap (although that one used to be rarer so more testing is needed).

However, I managed to reproduce the texture swap inside the Velvet room. In fact, this matches the behaviour I saw when I was using the old "full-range protection + "set_dirty(true)" workaround, where I could only reproduce texture swaps (rarely) in the velvet room. I suspect there might be a fourth bug in the tecture cache (or something related) that is only triggered inside the Velvet room.

Note that I was playing with "conservative protection", so it might be that this is a "conservative protection limitation" and "full-range protection" would resolve it. However, that was not the case with my previous work-around so more testing will be needed.

**Comment by ruipin** [comment]
#5013 has now been merged, meaning this should now be fixed on master. Let us know if you still see any texture swaps.

Otherwise (if nobody complains for a while), I think this bug can be closed.

**Comment by kd-11** [comment]
Closing as fixed via https://github.com/RPCS3/rpcs3/pull/5013
